src/lib/InstantSearch.ts

```diff
@@ -480,6 +483,20 @@ See ${createDocumentationLink({
       });
     });
 
+    if (this.state === 'disposed') {
+      // if the instance gets re-started,
+      // we re-create the middleware instances.
+      this.middleware = this.middleware.map(({ creator }) => ({
+        creator,
+        instance: {
+          subscribe: noop,
+          unsubscribe: noop,
+          onStateChange: noop,
+          ...creator({ instantSearchInstance: this }),
+        },
+      }));
+    }
+
     this.mainHelper = mainHelper;
 
     this.mainIndex.init({
```

Comment:
what if we move this to after the middleware subscribe happens? technically breaking behaviour, but I don't think it's something we defined as api (which order middleware happen in). Then we can move the initialUiState patch to before the init

Reply:
That's suggested in fix 2, and let me give it another try. It got me many failed test cases here and there, but I didn't looked at them closely. It could be trivial, or not.

Reply:
I guess the subscribe was expected to be called with a set ui state, which causes the tests to fail, although I'm not sure if that's really needed